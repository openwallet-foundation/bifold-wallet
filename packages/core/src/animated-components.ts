import ButtonLoading from './components/animated/ButtonLoading'
import ConnectionLoading from './components/animated/ConnectionLoading'
import CredentialAdded from './components/animated/CredentialAdded'
import CredentialPending from './components/animated/CredentialPending'
import LoadingIndicator from './components/animated/LoadingIndicator'
import RecordLoading from './components/animated/RecordLoading'
import SendingProof from './components/animated/SendingProof'
import SentProof from './components/animated/SentProof'

export interface AnimatedComponents {
  ButtonLoading: React.FC
  ConnectionLoading: React.FC
  CredentialAdded: React.FC
  CredentialPending: React.FC
  LoadingIndicator: React.FC
  RecordLoading: React.FC
  SendingProof: React.FC
  SentProof: React.FC
}

export const animatedComponents: AnimatedComponents = {
  ButtonLoading: ButtonLoading,
  ConnectionLoading: ConnectionLoading,
  CredentialAdded: CredentialAdded,
  CredentialPending: CredentialPending,
  LoadingIndicator: LoadingIndicator,
  RecordLoading: RecordLoading,
  SendingProof: SendingProof,
  SentProof: SentProof,
}
