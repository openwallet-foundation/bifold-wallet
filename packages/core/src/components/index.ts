import * as buttons from './buttons'
import * as misc from './misc/index'

export { misc, buttons }
