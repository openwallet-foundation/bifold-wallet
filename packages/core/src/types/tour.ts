import { ReactNode } from 'react'

/**
 * Makes the target optional. I.e, it can be the target type or `undefined`.
 */
export type Optional<T> = T | undefined

/**
 * An alias of what a React child looks when passed as function.
 */
export type ChildFn<T> = (value: T) => ReactNode

export enum BaseTourID {
  HomeTour = 'homeTourSteps',
  CredentialsTour = 'credentialsTourSteps',
  CredentialOfferTour = 'credentialOfferTourSteps',
  ProofRequestTour = 'proofRequestTourSteps',
}

export type TourID = `${string}TourSteps`
