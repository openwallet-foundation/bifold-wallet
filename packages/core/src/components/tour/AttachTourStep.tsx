import React, { cloneElement, ReactElement, ReactNode, RefObject, useContext, useEffect, useRef } from 'react'
import { StyleProp, useWindowDimensions, View } from 'react-native'

import { TourContext } from '../../contexts/tour/tour-context'
import { TourID } from '../../types/tour'
import { testIdWithKey } from '../../utils/testable'

export interface ChildProps<T> {
  children?: ReactNode
  ref: RefObject<unknown>
  style: StyleProp<T>
}

export interface AttachTourStepProps<T> {
  children: ReactElement<ChildProps<T>>
  /**
   * When `AttachTourStep` wraps a Functional Component, it needs to add an
   * aditional `View` on top of it to be able to measure the layout upon
   * render. This prop allows to define the behavior of the width of such
   * `View`. When set to `false`, it adjusts to its contents, when set to
   * `true`, it stretches out and tries to fill it view.
   *
   * **Note:** This prop has no effect when wrapping native components or
   * componentes created with `React.forwardRef`, which pass the `ref` to
   * another native component.
   *
   * @default false
   */
  fill?: boolean
  /**
   * The index of the `steps` array to which the step is attatched to.
   */
  index: number
  /**
   * The id of the tour this step belongs to (ie. home tour or proof request tour or etc.)
   */
  tourID: TourID
}

/**
 * React functional component used to attach and step to another component by
 * only wrapping it. Use its props to customize the behavior.
 *
 * @param props the component props
 * @returns an AttachTourStep React element
 */
export function AttachTourStep<T>({ children, fill = false, index, tourID }: AttachTourStepProps<T>): ReactElement {
  const { currentStep, currentTour, changeSpot } = useContext(TourContext)
  const { width: windowWidth, height: windowHeight } = useWindowDimensions()
  const childRef = useRef<View>(null)

  useEffect(() => {
    if (currentTour === tourID && currentStep === index) {
      childRef.current?.measureInWindow((x, y, width, height) => {
        changeSpot({ height, width, x, y })
      })
    }
  }, [currentTour, tourID, currentStep, index, windowWidth, windowHeight, changeSpot])

  const { style, ...rest } = children.props
  const childStyle = style ?? {}
  return (
    <View
      testID={testIdWithKey('AttachTourStep')}
      ref={childRef}
      style={{ alignSelf: fill ? 'stretch' : 'center', ...childStyle }}
      collapsable={false}
      focusable={false}
    >
      {cloneElement(children, rest, children.props?.children)}
    </View>
  )
}
