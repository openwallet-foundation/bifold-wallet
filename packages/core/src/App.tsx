import React, { useEffect } from 'react'
import { StatusBar } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import Toast from 'react-native-toast-message'

import { useNavigationContainerRef } from '@react-navigation/native'
import { isTablet } from 'react-native-device-info'
import Orientation from 'react-native-orientation-locker'
import { animatedComponents } from './animated-components'
import ErrorModal from './components/modals/ErrorModal'
import toastConfig from './components/toast/ToastConfig'
import { tours } from './constants'
import { Container, ContainerProvider } from './container-api'
import { AnimatedComponentsProvider } from './contexts/animated-components'
import { AuthProvider } from './contexts/auth'
import NavContainer from './contexts/navigation'
import { NetworkProvider } from './contexts/network'
import { StoreProvider } from './contexts/store'
import { ThemeProvider } from './contexts/theme'
import { TourProvider } from './contexts/tour/tour-provider'
import { initStoredLanguage } from './localization'
import RootStack from './navigators/RootStack'
import { bifoldTheme, themes } from './theme'

const createApp = (container: Container): React.FC => {
  const AppComponent: React.FC = () => {
    const navigationRef = useNavigationContainerRef()

    useEffect(() => {
      initStoredLanguage().then()
    }, [])

    useEffect(() => {
      // Hide the native splash / loading screen so that our
      // RN version can be displayed.
      SplashScreen.hide()
    }, [])

    if (!isTablet()) {
      Orientation.lockToPortrait()
    }

    return (
      <ContainerProvider value={container}>
        <StoreProvider>
          <ThemeProvider themes={themes} defaultThemeName={bifoldTheme.themeName}>
            <NavContainer navigationRef={navigationRef}>
              <AnimatedComponentsProvider value={animatedComponents}>
                <AuthProvider>
                  <NetworkProvider>
                    <StatusBar
                      hidden={false}
                      barStyle="light-content"
                      backgroundColor={bifoldTheme.ColorPalette.brand.primary}
                      translucent={false}
                    />
                    <ErrorModal />
                    <TourProvider tours={tours} overlayColor={'gray'} overlayOpacity={0.7}>
                      <RootStack />
                    </TourProvider>
                    <Toast topOffset={15} config={toastConfig} />
                  </NetworkProvider>
                </AuthProvider>
              </AnimatedComponentsProvider>
            </NavContainer>
          </ThemeProvider>
        </StoreProvider>
      </ContainerProvider>
    )
  }

  return AppComponent
}

export default createApp
