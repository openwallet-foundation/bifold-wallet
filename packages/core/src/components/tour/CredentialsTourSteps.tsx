import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { Text } from 'react-native'

import { useTheme } from '../../contexts/theme'
import { RenderProps, TourStep } from '../../contexts/tour/tour-context'

import { TourBox } from './TourBox'

export const credentialsTourSteps: TourStep[] = [
  {
    Render: (props: RenderProps) => {
      const { currentTour, currentStep, next, stop, previous } = props
      const { t } = useTranslation()
      const { ColorPallet, TextTheme } = useTheme()
      return (
        <TourBox
          title={t('Tour.AddCredentials')}
          hideLeft
          rightText={t('Tour.Done')}
          onRight={stop}
          currentTour={currentTour}
          currentStep={currentStep}
          previous={previous}
          stop={stop}
          next={next}
        >
          <Text
            style={{
              ...TextTheme.normal,
              color: ColorPallet.notification.infoText,
            }}
            allowFontScaling={false}
          >
            {t('Tour.AddCredentialsDescription')}
          </Text>
        </TourBox>
      )
    },
  },
]
