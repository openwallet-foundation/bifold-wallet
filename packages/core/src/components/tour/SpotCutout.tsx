import React, { useContext, useEffect, useRef } from 'react'
import { Animated, Easing } from 'react-native'
import { Rect } from 'react-native-svg'

import { TourContext } from '../../contexts/tour/tour-context'

const AnimatedRectangle = Animated.createAnimatedComponent(Rect)

export const SpotCutout = () => {
  const { spot } = useContext(TourContext)
  const spotPadding = 20

  const start = useRef(new Animated.ValueXY({ x: 0, y: 0 }))
  const width = useRef(new Animated.Value(0))
  const height = useRef(new Animated.Value(0))

  useEffect(() => {
    const { height: spotHeight, width: spotWidth, x, y } = spot
    const w = spotWidth + spotPadding
    const h = spotHeight + spotPadding
    const paddedX = x - spotPadding / 2
    const paddedY = y - spotPadding / 2

    const transition = (): Animated.CompositeAnimation => {
      return Animated.parallel([
        Animated.timing(start.current, {
          duration: 500,
          easing: Easing.out(Easing.exp),
          toValue: { x: paddedX, y: paddedY },
          useNativeDriver: false,
        }),
        Animated.timing(width.current, {
          duration: 500,
          easing: Easing.out(Easing.exp),
          toValue: w,
          useNativeDriver: false,
        }),
        Animated.timing(height.current, {
          duration: 500,
          easing: Easing.out(Easing.exp),
          toValue: h,
          useNativeDriver: false,
        }),
      ])
    }

    transition().start()
  }, [spot])

  if ([spot.height, spot.width].every((value) => value <= 0)) {
    return null
  }

  return (
    <AnimatedRectangle
      width={width.current}
      height={height.current}
      x={start.current.x}
      y={start.current.y}
      opacity={1}
      rx={10}
      fill="black"
    />
  )
}
