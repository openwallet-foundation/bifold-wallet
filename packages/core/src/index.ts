/* eslint-disable import/no-cycle */
import type { OnboardingStyleSheet } from './screens/Onboarding'

import { Agent } from '@credo-ts/core'
import AgentProvider from '@credo-ts/react-hooks'

import createApp from './App'
import * as components from './components'
import { ButtonImpl as Button, ButtonType, Button as IButton } from './components/buttons/Button'
import IconButton, { ButtonLocation } from './components/buttons/IconButton'
import BulletPoint from './components/inputs/BulletPoint'
import CheckBoxRow from './components/inputs/CheckBoxRow'
import LimitedTextInput from './components/inputs/LimitedTextInput'
import NotificationListItem from './components/listItems/NotificationListItem'
import ContentGradient from './components/misc/ContentGradient'
import CredentialCard from './components/misc/CredentialCard'
import ErrorBoundaryWrapper from './components/misc/ErrorBoundary'
import FauxHeader from './components/misc/FauxHeader'
import InfoBox, { InfoBoxType } from './components/misc/InfoBox'
import QRScannerTorch from './components/misc/QRScannerTorch'
import DeveloperModal from './components/modals/DeveloperModal'
import ErrorModal from './components/modals/ErrorModal'
import SafeAreaModal from './components/modals/SafeAreaModal'
import Record from './components/record/Record'
import InfoTextBox from './components/texts/InfoTextBox'
import Link from './components/texts/Link'
import Text from './components/texts/Text'
import { ThemedText } from './components/texts/ThemedText'
import { ToastType } from './components/toast/BaseToast'
import toastConfig from './components/toast/ToastConfig'
import { AttachTourStep } from './components/tour/AttachTourStep'
import { credentialOfferTourSteps } from './components/tour/CredentialOfferTourSteps'
import { credentialsTourSteps } from './components/tour/CredentialsTourSteps'
import { homeTourSteps } from './components/tour/HomeTourSteps'
import { proofRequestTourSteps } from './components/tour/ProofRequestTourSteps'
import { TourBox } from './components/tour/TourBox'
import { Banner, BannerMessage, BannerSection } from './components/views/Banner'
import HomeFooterView from './components/views/HomeFooterView'
import KeyboardView from './components/views/KeyboardView'
import { attemptLockoutConfig, PINRules, tours, walletTimeout } from './constants'
import { defaultConfig, defaultHistoryEventsLogger } from './container-impl'
import * as contexts from './contexts'
import { ActivityProvider, AutoLockTime, useActivity } from './contexts/activity'
import { AuthProvider } from './contexts/auth'
import NavContainer from './contexts/navigation'
import { NetworkProvider } from './contexts/network'
import { useTour } from './contexts/tour/tour-context'
import { TourProvider } from './contexts/tour/tour-provider'
import { useDeveloperMode } from './hooks/developer-mode'
import usePreventScreenCapture from './hooks/screen-capture'
import useBifoldAgentSetup from './hooks/useBifoldAgentSetup'
import { OpenIDCredentialRecordProvider } from './modules/openid/context/OpenIDCredentialRecordProvider'
import { DefaultScreenLayoutOptions } from './navigators/defaultLayoutOptions'
import { DefaultScreenOptionsDictionary, useDefaultStackOptions } from './navigators/defaultStackOptions'
import AttemptLockout from './screens/AttemptLockout'
import Biometry from './screens/Biometry'
import Developer from './screens/Developer'
import Onboarding from './screens/Onboarding'
import OnboardingPages from './screens/OnboardingPages'
import Preface from './screens/Preface'
import Scan from './screens/Scan'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import UpdateAvailable from './screens/UpdateAvailable'
import { bifoldLoggerInstance } from './services/bifoldLogger'
import { isBiometricsActive, loadLoginAttempt } from './services/keychain'
import { BifoldLogger } from './services/logger'
import { AbstractBifoldLogger } from './services/AbstractBifoldLogger'
import { DeepPartial, ThemeBuilder } from './theme-builder'
import * as types from './types'
import { CredentialListFooterProps } from './types/credential-list-footer'

export { animatedComponents } from './animated-components'
export { EventTypes, LocalStorageKeys } from './constants'
export { AnimatedComponentsProvider, useAnimatedComponents } from './contexts/animated-components'
export { useAuth } from './contexts/auth'
export { useNetwork } from './contexts/network'
export { DispatchAction, reducer, default as Store } from './contexts/reducers/store'
export { defaultState, mergeReducers, StoreContext, StoreProvider, useStore } from './contexts/store'
export { ThemeProvider, useTheme } from './contexts/theme'
export { useDeepLinks } from './hooks/deep-links'
export { initLanguages, initStoredLanguage, Locales, translationResources } from './localization'
export * from './navigators'
export { createStyles } from './screens/OnboardingPages'
export * from './services/storage'
export { bifoldTheme, ColorPalette, Assets as ImageAssets } from './theme'
export * from './types/attestation'
export { BifoldError } from './types/error'
export { Screens, Stacks, TabStacks } from './types/navigators'
export * from './types/version-check'
export { createLinkSecretIfRequired, getAgentModules } from './utils/agent'
export { getCredentialIdentifiers, isValidAnonCredsCredential } from './utils/credential'
export {
  connectFromScanOrDeepLink,
  formatTime,
  getConnectionName,
  removeExistingInvitationsById,
  useCredentialConnectionLabel,
} from './utils/helpers'
export { getIndyLedgers, IndyLedger, readIndyLedgersFromFile, writeIndyLedgersToFile } from './utils/ledger'
export { statusBarStyleForColor, StatusBarStyles } from './utils/luminance'
export { migrateToAskar } from './utils/migration'
export { buildFieldsFromAnonCredsCredential } from './utils/oca'
export { testIdForAccessabilityLabel, testIdWithKey } from './utils/testable'

export type { AnimatedComponents } from './animated-components'
export type { ReducerAction } from './contexts/reducers/store'
export type { RenderProps, TourStep } from './contexts/tour/tour-context'
export type {
  IAssets,
  IBrandColors,
  IColorPalette,
  IErrorColors,
  IFontAttributes,
  IGrayscaleColors,
  IInlineInputMessage,
  IInputAttributes,
  IInputs,
  INotificationColors,
  ISemanticColors,
  ISpacing,
  ISVGAssets,
  ITabTheme,
  ITextTheme,
  ITheme,
} from './theme'
export type { GenericFn } from './types/fn'
export { BasicMessageMetadata, CredentialMetadata } from './types/metadata'
export type { basicMessageCustomMetadata, credentialCustomMetadata } from './types/metadata'
export type { ContactStackParams, NotificationStackParams, OnboardingStackParams } from './types/navigators'
export type { WalletSecret } from './types/security'
export type {
  LoginAttempt as LoginAttemptState,
  Migration as MigrationState,
  Onboarding as OnboardingState,
  PersistentState,
  Preferences as PreferencesState,
  State,
  Tours as ToursState,
  VersionInfo,
} from './types/state'
export type { BifoldAgent } from './utils/agent'
export type { IndyLedgerConfig, IndyLedgerFileSystem, IndyLedgerJSON, IndyLedgersRecord } from './utils/ledger'
export type { OnboardingStyleSheet }

export type { InlineMessageProps } from './components/inputs/InlineErrorText'

export { InlineErrorPosition } from './types/error'

export * from './container-api'
export { MainContainer } from './container-impl'
export type { CredentialListFooterProps }

export type { Config, HistoryEventsLoggerConfig } from './types/config'
export type { OnboardingTask, ScreenLayoutConfig } from './types/navigators'

export { LockoutReason } from './contexts/auth'
export type { OnboardingStackProps } from './navigators/OnboardingStack'
export type { SplashProps } from './screens/Splash'
export { BaseTourID } from './types/tour'

export type { BannerSectionProps } from './components/views/Banner'

export {
  ActivityProvider,
  Agent,
  AgentProvider,
  AttachTourStep,
  AttemptLockout,
  attemptLockoutConfig,
  AuthProvider,
  AutoLockTime,
  Banner,
  BannerSection,
  AbstractBifoldLogger,
  BifoldLogger,
  bifoldLoggerInstance,
  Biometry,
  BulletPoint,
  Button,
  ButtonLocation,
  ButtonType,
  CheckBoxRow,
  components,
  ContentGradient,
  contexts,
  createApp,
  CredentialCard,
  credentialOfferTourSteps,
  credentialsTourSteps,
  defaultConfig,
  defaultHistoryEventsLogger,
  DefaultScreenLayoutOptions,
  DefaultScreenOptionsDictionary,
  Developer,
  DeveloperModal,
  ErrorBoundaryWrapper,
  ErrorModal,
  FauxHeader,
  HomeFooterView as HomeContentView,
  homeTourSteps,
  IconButton,
  InfoBox,
  InfoBoxType,
  InfoTextBox,
  isBiometricsActive,
  KeyboardView,
  LimitedTextInput,
  Link,
  loadLoginAttempt,
  NavContainer,
  NetworkProvider,
  NotificationListItem,
  Onboarding,
  OnboardingPages,
  OpenIDCredentialRecordProvider,
  PINRules,
  Preface,
  proofRequestTourSteps,
  QRScannerTorch,
  Record,
  SafeAreaModal,
  Scan,
  Splash,
  Terms,
  Text,
  ThemeBuilder,
  ThemedText,
  toastConfig,
  ToastType,
  TourBox,
  TourProvider,
  tours,
  types,
  UpdateAvailable,
  useActivity,
  useBifoldAgentSetup,
  useDefaultStackOptions,
  useDeveloperMode,
  usePreventScreenCapture,
  useTour,
  walletTimeout,
}
export type { BannerMessage, DeepPartial, IButton }
