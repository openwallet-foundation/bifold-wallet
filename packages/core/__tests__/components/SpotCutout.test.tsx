import { act, render } from '@testing-library/react-native'
import React from 'react'

import { SpotCutout } from '../../src/components/tour/SpotCutout'
import { TourProvider } from '../../src/contexts/tour/tour-provider'
import { tours } from '../../src/constants'

describe('SpotCutout Component', () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })
  afterAll(() => {
    jest.useRealTimers()
  })
  test('Renders properly with defaults', async () => {
    const tree = render(
      <TourProvider tours={tours} overlayColor={'gray'} overlayOpacity={0.7}>
        <SpotCutout />
      </TourProvider>
    )
    await act(() => {
      jest.runAllTimers()
    })
    expect(tree).toMatchSnapshot()
  })
})
