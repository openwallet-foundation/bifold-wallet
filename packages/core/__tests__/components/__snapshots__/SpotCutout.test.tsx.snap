// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotCutout Component Renders properly with defaults 1`] = `null`;
